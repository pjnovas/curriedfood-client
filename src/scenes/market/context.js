import React from 'react';
import { get } from 'lodash';

import { useRequest, useRequestSWR } from 'hooks/service';

const ShoppingContext = React.createContext();

const getUri = (id) => `/shopping-lists/${id || ''}`;

const useUpdate = ({ data, revalidate }) => {
  const request = useRequest();

  return async (groceries) => {
    await request({
      url: getUri(data.id),
      method: 'put',
      data: {
        groceries
      }
    });

    revalidate();
  };
};

const ShoppingProvider = ({ children }) => {
  const value = useRequestSWR({ url: getUri() }, { withPlace: true });
  const data = get(value, 'data[0]') || null;
  const update = useUpdate({ ...value, data });

  const parsed = {
    ...value,
    update,
    data,
    mutate: (groceries) => {
      value.mutate({
        ...data,
        groceries
      });
    }
  };

  return (
    <ShoppingContext.Provider value={parsed}>
      {children}
    </ShoppingContext.Provider>
  );
};

const useShopping = () => {
  const context = React.useContext(ShoppingContext);

  if (context === undefined) {
    throw new Error(`useShopping must be used within a ShoppingProvider`);
  }

  return context;
};

export { ShoppingProvider, useShopping };
