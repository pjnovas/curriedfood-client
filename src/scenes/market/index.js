export { ShoppingListScreen } from './shopping-list';
export { IngredientEdit } from './ingredient-edit';
export { IngredientNew } from './ingredient-new';
