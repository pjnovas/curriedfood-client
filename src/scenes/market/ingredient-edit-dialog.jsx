import React, { useState } from 'react';
import { StyleSheet } from 'react-native';
import { Portal, Dialog, Button } from 'react-native-paper';
import get from 'lodash/get';
import Theme from 'theme';
import Chips from 'components/chips';

export const IngredientEdit = ({ onDismiss, onSubmit, allTags, ...item }) => {
  const [tags, setValue] = useState(get(item, 'product.shop_tags', ''));

  return (
    <Portal>
      <Dialog style={styles.layout} visible dismissable onDismiss={onDismiss}>
        <Dialog.Title style={styles.title}>{item.product.name}</Dialog.Title>
        <Dialog.Content style={styles.content}>
          <Chips
            list={allTags}
            value={tags}
            onChange={(value) => {
              setValue(value);
              console.log(value);
            }}
          />
        </Dialog.Content>
        <Dialog.Actions style={styles.actions}>
          <Button onPress={onDismiss}>Cancelar</Button>
          <Button
            dark
            mode="contained"
            style={styles.okButton}
            onPress={() => onSubmit(tags)}
          >
            OK
          </Button>
        </Dialog.Actions>
      </Dialog>
    </Portal>
  );
};

const styles = StyleSheet.create({
  layout: {},
  title: {
    color: Theme.colors.text,
    textAlign: 'center'
  },
  content: {
    minHeight: 100,
    maxHeight: 400
  },
  actions: {},
  okButton: {
    marginHorizontal: 10
  }
});

// export const useFormMarket = () => ({
//   form: useForm()
// });

// export default composeHooks({ useFormMarket })(IngredientEdit);
export default IngredientEdit;
