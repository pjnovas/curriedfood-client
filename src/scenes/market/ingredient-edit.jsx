import React from 'react';
import { StyleSheet } from 'react-native';
import { Formik } from 'formik';
import { FAB } from 'react-native-paper';

import find from 'lodash/find';
import get from 'lodash/get';

import { useRequestSWR } from 'hooks/service';
import { useNavigateBack } from 'hooks/navigation';
import { SafeAreaLayout, SaveAreaInset } from 'components/safe-area-layout';

import { FormInput } from 'components/form-input';
import { Toolbar } from 'components/toolbar';
import Spinner from 'components/spinner';
import Layout from 'components/layout';

// TODO: should check cache and fetch by id (web mode)

export const IngredientEdit = ({
  route: {
    params: { id, title }
  }
}) => {
  const { data } = useRequestSWR(
    {
      url: `/shopping-lists`
    },
    { withPlace: true }
  );

  const item = data && find(get(data, '[0].groceries'), { id });

  const renderForm = ({ handleSubmit }) => (
    <>
      <FormInput id="quantity" style={styles.formControl} label="Cantidad" />
      <FAB style={styles.fab} icon="save" onPress={handleSubmit} />
    </>
  );

  return (
    <SafeAreaLayout style={styles.safeArea} insets={SaveAreaInset.TOP}>
      <Toolbar title={`Editar ${title}`} onBackPress={useNavigateBack()} />
      {!item ? (
        <Spinner bar />
      ) : (
        <Layout style={styles.form}>
          <Formik
            initialValues={item}
            // validationSchema={}
            onSubmit={(values) => console.log(values)}
          >
            {renderForm}
          </Formik>
        </Layout>
      )}
    </SafeAreaLayout>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    paddingBottom: 0,
    position: 'relative'
  },
  form: {
    padding: 20
  },
  formContainer: {
    flex: 1,
    paddingVertical: 16,
    paddingHorizontal: 16
  },
  fab: {
    position: 'absolute',
    margin: 16,
    right: 0,
    bottom: 0
  }
});

export default IngredientEdit;
