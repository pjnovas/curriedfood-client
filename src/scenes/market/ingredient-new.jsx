import React from 'react';
import { Text, StyleSheet } from 'react-native';
import { useNavigateBack } from 'hooks/navigation';

import { SafeAreaLayout, SaveAreaInset } from 'components/safe-area-layout';
import { Toolbar } from 'components/toolbar';
import Layout from 'components/layout';

export const IngredientNew = () => (
  <SafeAreaLayout style={styles.safeArea} insets={SaveAreaInset.TOP}>
    <Toolbar title="Nuevo Ingrediente" onBackPress={useNavigateBack()} />
    <Layout>
      <Text>Nuevo Ingrediente</Text>
    </Layout>
  </SafeAreaLayout>
);

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    paddingBottom: 0,
    position: 'relative'
  }
});

export default IngredientNew;
