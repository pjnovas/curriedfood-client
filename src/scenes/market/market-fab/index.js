import FabButton from './fab-button';

export default FabButton;
