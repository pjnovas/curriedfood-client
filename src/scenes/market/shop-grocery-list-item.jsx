import React from 'react';
import { ListItem, Text } from '@ui-kitten/components';

const unitText = {
  un: '',
  cu: 'cucharadas de',
  gr: 'gramos de',
  lt: 'litros de',
  tz: 'tazas de',
  ml: 'mililitros'
};

const getText = ({
  product: { name },
  quantity,
  unit: { code },
  alt_quantity,
  alt_unit
}) => {
  let text = `${quantity} ${unitText[code] || ''} ${name}`;

  if (alt_quantity) {
    text += ` (${alt_quantity} ${alt_unit.code})`;
  }

  return text;
};

const ShopGroceryListItem = (props) => (
  <ListItem {...props}>
    <Text category="s1">{getText(props)}</Text>
  </ListItem>
);

export default ShopGroceryListItem;
