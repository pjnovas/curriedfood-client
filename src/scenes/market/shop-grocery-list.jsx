import React, { Fragment, useState, useEffect } from 'react';
import { ScrollView, StyleSheet } from 'react-native';
import { FAB, List, TextInput, IconButton, Divider } from 'react-native-paper';
import { isEmpty, includes, uniq, compact, sortBy, mapValues } from 'lodash';

import { AppRoute } from 'navigation/app-routes';
import { useNavigateTo } from 'hooks/navigation';
import { longName, convert } from 'utils/grocery';
import { composeHooks } from 'utils/language';
import ProductTagsDialog from 'scenes/products/product-tags-dialog';

export const ShopGroceryList = ({
  showEditItem,
  openNewItem,
  reload,
  closeEdit,
  openEdition,
  categories
}) => (
  <>
    <ScrollView contentContainerStyle={styles.layout}>
      {Object.keys(categories).map((key) => (
        <List.Section key={key}>
          <List.Subheader style={styles.sectionTitle}>{key}</List.Subheader>
          {categories[key].map((item) => (
            <Fragment key={item.id}>
              <List.Item
                key={item.id}
                title={item.product.name}
                description={longName[item.product.unit]}
                onPress={openEdition(item.product)}
                left={(props) => (
                  <TextInput
                    {...props}
                    selectTextOnFocus
                    style={styles.quantity}
                    defaultValue={convert(item.quantity)}
                    mode="outlined"
                    keyboardType="numeric"
                    textAlign="right"
                    maxLength={20}
                    dense
                  />
                )}
                right={(props) => <IconButton {...props} icon="close" />}
              />
              <Divider />
            </Fragment>
          ))}
        </List.Section>
      ))}
    </ScrollView>

    <FAB style={styles.fab} icon="plus" onPress={openNewItem} />
    {showEditItem && (
      <ProductTagsDialog
        tagField="shop_tags"
        onDismiss={closeEdit}
        onSubmit={() => {
          reload();
          closeEdit();
        }}
        {...showEditItem}
      />
    )}
  </>
);

const styles = StyleSheet.create({
  layout: {
    paddingBottom: 80
  },
  sectionTitle: {
    fontSize: 16,
    textTransform: 'uppercase',
    alignSelf: 'center'
  },
  quantity: {
    width: 80
  },
  fab: {
    position: 'absolute',
    margin: 16,
    right: 0,
    bottom: 0
  }
});

export const useNavigation = ({ data }) => {
  const newIngredient = useNavigateTo(AppRoute.GROCERIES_NEW);
  const [showEditItem, setVisibleEdit] = useState();
  const [categories, setCategories] = useState({});

  useEffect(() => {
    const tags = compact(
      uniq(
        data.reduce(
          (all, item) => [
            ...all,
            ...(item?.product.shop_tags || '').split(',')
          ],
          []
        )
      )
    );

    tags.sort();

    let resolved = tags.reduce(
      (all, tag) => ({
        ...all,
        [tag]: data.filter((item) => includes(item.product.shop_tags, tag))
      }),
      {
        'sin categoria': data.filter((item) => isEmpty(item.product.shop_tags))
      }
    );

    setCategories(
      mapValues(resolved, (item) => sortBy(item, ({ product }) => product.name))
    );
  }, [data]);

  return {
    openNewItem: () => newIngredient(),
    openEdition: (product) => () =>
      setVisibleEdit({
        product,
        allTags:
          Object.keys(categories).filter((key) => key !== 'sin categoria') || []
      }),
    closeEdit: () => setVisibleEdit(),
    reload: () => {
      // TODO: re fetch to remap categories
    },
    showEditItem,
    categories
  };
};

export default composeHooks({ useNavigation })(ShopGroceryList);
