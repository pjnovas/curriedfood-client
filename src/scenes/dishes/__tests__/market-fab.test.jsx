import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import { FAB } from 'react-native-paper';

import { useMarket } from 'hooks/market';
import MarketFab, { ConfirmDialog } from '../market-fab';

jest.mock('hooks/navigation');
jest.mock('hooks/market', () => ({
  useMarket: jest.fn()
}));

const props = {
  ingredients: [
    {
      id: 1,
      quantity: 3,
      product: {
        name: 'Eggs',
        divisible: false
      },
      unit: {
        code: 'un'
      }
    },
    {
      id: 2,
      quantity: 150,
      product: {
        name: 'Olives',
        divisible: true
      },
      unit: {
        code: 'gr'
      }
    }
  ],
  servings: 4
};

describe('MarketFab', () => {
  takeSnapshots(
    {
      'renders nothing with no ingredients': {
        ...props,
        ingredients: []
      },
      'renders a FAB Button with ingredients': props
    },
    MarketFab
  );

  test('renders a ConfirmDialog when FAB is clicked', () => {
    const wrapper = shallow(<MarketFab {...props} />);
    wrapper.find(FAB).simulate('press');
    takeSnapshotFrom(wrapper);
  });

  test('renders a FAB Button when confirm dialog is dismissed', () => {
    const wrapper = shallow(<MarketFab {...props} />);
    wrapper.find(FAB).simulate('press'); // open dialog
    wrapper.find(ConfirmDialog).simulate('dismiss');
    takeSnapshotFrom(wrapper);
  });
});

describe('ConfirmDialog', () => {
  useMarket.mockImplementation(() => [{}, { mergeList: jest.fn() }]);

  takeSnapshots(
    {
      default: props
    },
    ConfirmDialog
  );
});
