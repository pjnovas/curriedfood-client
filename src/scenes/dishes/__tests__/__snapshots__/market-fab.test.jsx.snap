// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog default 1`] = `
<Confirmation
  loading={true}
  okText="Enviar"
  onConfirm={[Function]}
  title="Enviar al carrito de compras?"
  visible={true}
>
  <Paragraph
    style={
      Object {
        "color": "#ffffff",
        "textAlign": "center",
      }
    }
  >
    Se agregar√°n 2 Ingredientes
  </Paragraph>
</Confirmation>
`;

exports[`MarketFab renders a ConfirmDialog when FAB is clicked 1`] = `
<ConfirmDialog
  ingredients={
    Array [
      Object {
        "id": 1,
        "product": Object {
          "divisible": false,
          "name": "Eggs",
        },
        "quantity": 3,
        "unit": Object {
          "code": "un",
        },
      },
      Object {
        "id": 2,
        "product": Object {
          "divisible": true,
          "name": "Olives",
        },
        "quantity": 150,
        "unit": Object {
          "code": "gr",
        },
      },
    ]
  }
  onDismiss={[Function]}
/>
`;

exports[`MarketFab renders a FAB Button when confirm dialog is dismissed 1`] = `
<withTheme(FAB)
  disabled={false}
  icon="cart-plus"
  onPress={[Function]}
  style={
    Object {
      "bottom": 0,
      "margin": 16,
      "position": "absolute",
      "right": 0,
    }
  }
/>
`;

exports[`MarketFab renders a FAB Button with ingredients 1`] = `
<withTheme(FAB)
  disabled={false}
  icon="cart-plus"
  onPress={[Function]}
  style={
    Object {
      "bottom": 0,
      "margin": 16,
      "position": "absolute",
      "right": 0,
    }
  }
/>
`;

exports[`MarketFab renders nothing with no ingredients 1`] = `""`;
